mangaki/mangaki/utils/manami.py

```diff
@@ -0,0 +1,522 @@
+            self._raw[pk]['nb_episodes'] = self._raw[pk]['nb_episodes'].split('x')[0]
+            category = category_from_type[self._raw[pk]['anime_type']]
+            self._raw[pk]['subcategory'] = coarse_category[category.lower()] if category else None
+
+        for work_id, synonym in WorkTitle.objects.values_list('work_id', 'title'):
+            while work_id in redirects:
+                work_id = redirects[work_id]
+            if work_id not in self._raw:
+                continue
+
+            self._raw[work_id].setdefault('synonyms', []).append(synonym)
+            self.from_synonym.setdefault(sanitize(synonym), set()).add(work_id)
+
+        self.df = pd.DataFrame.from_dict(self._raw, orient='index')
+
+    def __getitem__(self, key):
+        return self._raw[key]
+
+    def __len__(self):
+        return len(self._raw)
+
+    def print_summary(self):
+        if self._filter_sources is None:
+            filters = ''
+        else:
+            filters = '{} '.format(', '.join(self._filter_sources))
+
+        print('Mangaki:')
+        print(f'    {len(self)} animes')
+        print(f'    {len(self.references)} unique {filters}references')
+        if self.missing_refs:
+            print(f'    {len(self.missing_refs)} animes with no {filters}references')
+
+
+def get_clusters_from_ref(manami, dead, manami_cluster_backup=None):
+    references = set()
+    url1 = defaultdict(list)
+    url2 = defaultdict(set)
+
+    # Get all sources of Manami entries
+    for manami_id, entry in enumerate(manami):
+        references.add(('Manami', str(manami_id)))
+        for ref in entry['references']:
+            assert check_ref(dead, ref)
+            references.add(ref)
+            url1[manami_id].append(ref)
+
+    # Manami extra refs
+    if manami_cluster_backup:
+        for manami_cluster in manami_cluster_backup:
+            for url_cluster in manami_cluster:
+                for url in url_cluster:
+                    ref = parse_manami_source(url)
+                    references.add(ref)
+
+    # Mangaki works with AniDB ID
+    anidb_id = dict(Work.objects.filter(anidb_aid__gt=0).values_list('id', 'anidb_aid'))
+    for mangaki_id, aid in anidb_id.items():
+        if str(aid) not in dead['AniDB']:
+            ref1 = ('Mangaki', str(mangaki_id))
+            ref2 = ('AniDB', str(aid))
+            references.update([ref1, ref2])
+            url2[mangaki_id].update([ref1, ref2])
+
+    # All Mangaki references
+    for mangaki_id, source, identifier in Reference.objects.filter(work_id__category__slug='anime').values_list('work_id', 'source', 'identifier'):
+        ref1 = ('Mangaki', str(mangaki_id))
+        ref2 = (source, identifier)
+        if check_ref(dead, ref2):
+            references.update([ref1, ref2])
+            url2[mangaki_id].update([ref1, ref2])
+
+    ids = dict(zip(sorted(references), range(len(references))))
```

Comment:
j'avoue que ça manque trop de types pour que j'arrive à suivre ce qui se passe
tu renumérotes dans l'ordre induit pour ton objet `references`, mais je ne sais pas pourquoi tu fais ça

Reply:
Je veux des ID uniques pour les références

Reply:
oui mais pourquoi sur les références *triés* par l'ordre induit? pourquoi pas `dict(zip(references), range(len(references))` par exemple?

Reply:
ou genre `{v: k for k ,v in enumerate(references)}` tu vois aussi, ce genre de choses?

Reply:
J'ai repris ton snippet mais j'ai gardé le sorted ; il n'y a aucune raison °mais° ça m'a été utile lors du debug (de remarquer que j'avais beaucoup de références à chaîne vide, par exemple)